rlmolecule/alphazero/alphazero_problem.py

```diff
@@ -62,8 +63,14 @@ def reward_wrapper(self, state: GraphSearchState) -> Reward:
                                  state=state.serialize(),
                                  reward=reward,
                                  data=data)
-            self.session.merge(record)
-            self.session.commit()
+
+            try:
+                # Here we handle the possibility of a race condition where another thread has already added the
+                # current state's reward to the database before this thread has completed.
+                self.session.merge(record)
+                self.session.commit()
+            except exc.IntegrityError:
+                self.session.rollback()
```

Comment:
Would we want to log some kind of warning here?

Reply:
seems reasonable